src/snowflake/connector/cursor.py

```diff
@@ -144,20 +144,221 @@ class ResultMetadata(NamedTuple):
     def from_column(cls, col: dict[str, Any]):
         """Initializes a ResultMetadata object from the column description in the query response."""
+        type_code = FIELD_NAME_TO_ID[
+            col["extTypeName"].upper()
+            if col.get("extTypeName")
+            else col["type"].upper()
+        ]
+
         return cls(
             col["name"],
+            type_code,
+            None,
+            col["length"],
+            col["precision"],
+            col["scale"],
+            col["nullable"],
+        )
+
+
+class ResultMetadataV2Field:
+    """ResultMetadataV2Field represents the type information of one sub-type in a nested type."""
+
+    def __init__(
+        self,
+        type_code: int,
+        is_nullable: bool,
+        internal_size: int | None = None,
+        precision: int | None = None,
+        scale: int | None = None,
+        fields: list[ResultMetadataV2] | None = None,
+    ):
+        self._type_code = type_code
+        self._is_nullable = is_nullable
+        self._internal_size = internal_size
+        self._precision = precision
+        self._scale = scale
+        self._fields = fields
+
+    @classmethod
+    def from_column(cls, col: dict[str, Any]) -> ResultMetadataV2Field:
+        """Initializes a ResultMetadataV2Field object from a child of the column description in the query response."""
+        fields = None
+        if col.get("fields") is not None:
+            fields = [cls.from_column(f) for f in col["fields"]]
+        return cls(
             FIELD_NAME_TO_ID[
                 col["extTypeName"].upper()
                 if col.get("extTypeName")
                 else col["type"].upper()
             ],
-            None,
+            col["nullable"],
             col["length"],
             col["precision"],
             col["scale"],
+            fields,
+        )
+
+    @property
+    def type_code(self) -> int:
+        return self._type_code
+
+    @property
+    def is_nullable(self) -> bool:
+        return self._is_nullable
+
+    @property
+    def internal_size(self) -> int | None:
+        return self._internal_size
+
+    @property
+    def precision(self) -> int | None:
+        return self._precision
+
+    @property
+    def scale(self) -> int | None:
+        return self._scale
+
+    @property
+    def fields(self) -> list[ResultMetadataV2Field] | None:
+        return self._fields
+
+
+class ResultMetadataV2:
+    """ResultMetadataV2Field represents the type information of a single column.
+
+    It is a replacement for ResultMetadata that contains additional attributes, currently
+    `vector_dimension` and `fields`. This class will be unified with ResultMetadata in the
+    near future.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        type_code: int,
+        is_nullable: bool,
+        display_size: int | None = None,
+        internal_size: int | None = None,
+        precision: int | None = None,
+        scale: int | None = None,
```

Comment:
is there any particular reason why is_nullable moved up in order here compared to ResultMetadata. Does the ordering even matter?

Reply:
ordering does not matter. changes look good

Reply:
Yeah, I moved it up only since it doesn't have a default value. `ResultMetadataV2` isn't a named tuple, so it shouldn't have any other user-visible effect.